core/server/proxy/src/main/java/alluxio/proxy/s3/S3Handler.java

```diff
@@ -0,0 +1,338 @@
+            rejectUnsupportedResources();
+            // Init utils
+            mMetaFS = ProxyWebServer.mFileSystem;
+            mAsyncAuditLogWriter = ProxyWebServer.mAsyncAuditLogWriter;
+            // Initiate the S3 API metadata directories -> TODO this should be init once instead every time
+            if (!mMetaFS.exists(new AlluxioURI(S3RestUtils.MULTIPART_UPLOADS_METADATA_DIR))) {
+                mMetaFS.createDirectory(
+                        new AlluxioURI(S3RestUtils.MULTIPART_UPLOADS_METADATA_DIR),
+                        CreateDirectoryPOptions.newBuilder()
+                                .setRecursive(true)
+                                .setMode(PMode.newBuilder()
+                                        .setOwnerBits(Bits.ALL)
+                                        .setGroupBits(Bits.ALL)
+                                        .setOtherBits(Bits.NONE).build())
+                                .setWriteType(S3RestUtils.getS3WriteType())
+                                .setXattrPropStrat(XAttrPropagationStrategy.LEAF_NODE)
+                                .build()
+                );
+            }
+        } catch (Exception ex) {
+            LOG.error(ThreadUtils.formatStackTrace(ex));
+            throw S3RestUtils.toBucketS3Exception(ex, mBucket);
+        }
+    }
+
+    public void setS3Task(S3BaseTask task) { mS3Task = task; }
+
+    public S3BaseTask getS3Task() {
+        return mS3Task;
+    }
+
+    public static S3Handler createHandler(String path,
+                                       HttpServletRequest request,
+                                       HttpServletResponse response) throws S3Exception {
+        Stopwatch stopwatch = Stopwatch.createStarted();
+        Matcher baseMatcher = mBasePathPattern.matcher(path);
+        Matcher bucketMatcher = mBucketPathPattern.matcher(path);
+        Matcher objectMatcher = mObjectPathPattern.matcher(path);
+
+        String pathStr = path;
+        String bucket = null;
+        String object = null;
+        S3Handler handler = null;
+        try {
+            if (bucketMatcher.matches()) {
+                pathStr = path.substring(S3RequestServlet.S3_SERVICE_PATH_PREFIX.length() + 1);
+                bucket = URLDecoder.decode(pathStr, "UTF-8");
+            } else if (objectMatcher.matches()) {
+                pathStr = path.substring(S3RequestServlet.S3_SERVICE_PATH_PREFIX.length() + 1);
+                bucket = URLDecoder.decode(pathStr.substring(0, pathStr.indexOf(AlluxioURI.SEPARATOR)), "UTF-8");
+                object = URLDecoder.decode(pathStr.substring(pathStr.indexOf(AlluxioURI.SEPARATOR) + 1), "UTF-8");
+            }
+            handler = new S3Handler(bucket, object, request, response);
+            handler.setStopwatch(stopwatch);
+            handler.init();
+            S3BaseTask task = null;
+            if (object != null && !object.isEmpty()) {
+                task = S3ObjectTask.allocateTask(handler);
```

Comment:
I would prefer having a Factory class here for this thing, but it's not a big deal.

Reply:
i can put it into a factory style to have a uniform format with the others